x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -418,27 +476,114 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
-                final JwtRealm.JwksAlgs jwksAndAlgs = isJwtAlgHmac ? this.jwksAlgsHmac : this.jwksAlgsPkc;
-                JwtValidateUtil.validate(
-                    jwt,
-                    this.allowedIssuer,
-                    this.allowedAudiences,
-                    this.allowedClockSkew.seconds(),
-                    jwksAndAlgs.algs,
-                    jwksAndAlgs.jwks
-                );
+                header = jwt.getHeader();
                 claimsSet = jwt.getJWTClaimsSet();
-                LOGGER.trace("Realm [{}] JWT validation succeeded for token=[{}].", super.name(), tokenPrincipal);
+                final Date now = new Date();
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(
+                        "Realm [{}] JWT parse succeeded for token=[{}]."
+                            + "Validating JWT, now [{}], alg [{}], issuer [{}], audiences [{}], kty [{}],"
+                            + " auth_time [{}], iat [{}], nbf [{}], exp [{}], kid [{}], jti [{}]",
+                        super.name(),
+                        tokenPrincipal,
+                        now,
+                        header.getAlgorithm(),
+                        claimsSet.getIssuer(),
+                        claimsSet.getAudience(),
+                        header.getType(),
+                        claimsSet.getDateClaim("auth_time"),
+                        claimsSet.getIssueTime(),
+                        claimsSet.getNotBeforeTime(),
+                        claimsSet.getExpirationTime(),
+                        header.getKeyID(),
+                        claimsSet.getJWTID()
+                    );
+                }
+                // Validate all else before signature, because these checks are more helpful diagnostics than rejected signatures.
+                final boolean isJwtSigPkc = JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_PKC.contains(header.getAlgorithm().getName());
+                if (isJwtSigPkc && (this.isConfiguredJwkSetPkc == false)) {
+                    final String msg = "Realm [" + super.name() + "] not configured to verify PKC JWTs for token=[" + tokenPrincipal + "].";
+                    LOGGER.debug(msg);
+                    listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                    return;
+                } else if ((isJwtSigPkc == false) && (this.isConfiguredJwkSetHmac == false) && (this.isConfiguredJwkOidcHmac == false)) {
+                    final String msg = "Realm ["
+                        + super.name()
+                        + "] not configured to verify HMAC JWTs for token=["
+                        + tokenPrincipal
+                        + "].";
+                    LOGGER.debug(msg);
+                    listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                    return;
+                }
```

Comment:
They are similar checks, but not quite the same.

Example 1: Realm algs=[RS256], JWT alg=HS256 => Diagnostic exception, realm does not support any HMAC algorithms
Example 2: Realm algs=[HS256], JWT alg=HS512 => validateSignatureAlgorithm exception, realm supports some HMAC algorithms, but not the requested algorithm

I had a comment in the code that it was for diagnotics. However, the value of those diagnostic messages are limited because validateSignatureAlgorithm will catch it later on.

I think this code can be removed, as the added diagnostic value does not seem worth it to know if a realm has none or some of same HMAC or PKC algorithms.

Reply:
Done